plugins/vpp/abfplugin/descriptor/abf.go

```diff
@@ -0,0 +1,281 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package descriptor
+
+import (
+	"strconv"
+
+	"github.com/go-errors/errors"
+	"github.com/gogo/protobuf/proto"
+	prototypes "github.com/gogo/protobuf/types"
+	"github.com/ligato/cn-infra/idxmap"
+	"github.com/ligato/cn-infra/logging"
+	abf "github.com/ligato/vpp-agent/api/models/vpp/abf"
+	acl "github.com/ligato/vpp-agent/api/models/vpp/acl"
+	"github.com/ligato/vpp-agent/api/models/vpp/interfaces"
+	"github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	"github.com/ligato/vpp-agent/plugins/vpp/abfplugin/abfidx"
+	"github.com/ligato/vpp-agent/plugins/vpp/abfplugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/vpp/abfplugin/vppcalls"
+	"github.com/ligato/vpp-agent/plugins/vpp/aclplugin/aclidx"
+	ifdescriptor "github.com/ligato/vpp-agent/plugins/vpp/ifplugin/descriptor"
+)
+
+const (
+	// ABFDescriptorName is descriptor name
+	ABFDescriptorName = "vpp-abf"
+
+	// dependency labels
+	aclDep = "acl-exists"
+)
+
+// A list of non-retriable errors:
+var (
+	// ErrABFInvalidIndex is returned when ABF configuration is defined with invalid index (not a number).
+	ErrABFInvalidIndex = errors.New("ABF configuration contains invalid index")
+
+	// ErrABFWithoutACL is returned when ABF configuration does not contain associated access list.
+	ErrABFWithoutACL = errors.New("ABF configuration defined without ACL")
+)
+
+// ABFDescriptor is descriptor for ABF
+type ABFDescriptor struct {
+	// dependencies
+	log        logging.Logger
+	abfHandler vppcalls.ABFVppAPI
+
+	// runtime
+	aclIndex aclidx.ACLMetadataIndex
+}
+
+// NewABFDescriptor is constructor for ABF descriptor
+func NewABFDescriptor(abfHandler vppcalls.ABFVppAPI, aclIndex aclidx.ACLMetadataIndex,
+	logger logging.PluginLogger) *ABFDescriptor {
+	return &ABFDescriptor{
+		log:        logger.NewLogger("abf-descriptor"),
+		aclIndex:   aclIndex,
+		abfHandler: abfHandler,
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with
+// the KVScheduler.
+func (d *ABFDescriptor) GetDescriptor() *adapter.ABFDescriptor {
+	return &adapter.ABFDescriptor{
+		Name:          ABFDescriptorName,
+		NBKeyPrefix:   abf.ModelABF.KeyPrefix(),
+		ValueTypeName: abf.ModelABF.ProtoName(),
+		KeySelector:   abf.ModelABF.IsKeyValid,
+		KeyLabel:      abf.ModelABF.StripKeyPrefix,
+		WithMetadata:  true,
+		MetadataMapFactory: func() idxmap.NamedMappingRW {
+			return abfidx.NewABFIndex(d.log, "vpp-abf-index")
+		},
+		ValueComparator:      d.EquivalentABFs,
+		Validate:             d.Validate,
+		Create:               d.Create,
+		Delete:               d.Delete,
+		UpdateWithRecreate:   d.UpdateWithRecreate,
+		Retrieve:             d.Retrieve,
+		IsRetriableFailure:   d.IsRetriableFailure,
+		DerivedValues:        d.DerivedValues,
+		Dependencies:         d.Dependencies,
+		RetrieveDependencies: []string{ifdescriptor.InterfaceDescriptorName},
+	}
```

Comment:
Add ACL descriptor into RetrieveDependencies - aclIndex must be up-to-date for the retrieval of ABF policies.